### PR TITLE
DAOS-13216 dfuse: Add a pre-read feature for non-cached files.

### PR DESCRIPTION
**Pull request by ashleypittman**
When the kernel cache is in use but a file is not cached
then pre-read the file on open.

Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by mchaarawi** [comment]
have not dug into the code yet in this PR and have just some high level questions based on the comment of pre-read you added:
- if we are doing this ls -R or recursive readdir like resnet does initially from every thread before reading any data, are we going to trigger this pre-read? (that would be disastrous). i don't think the file is opened in this case, but not sure.
- in the case of a pre-read during the open, do we read 1 mib or up to the size of the file? is this querying the file size, or does fuse or dfuse already have queried the file size before issuing the open?
- is there a practical application this is aimed for?

**Comment by ashleypittman** [comment]
> have not dug into the code yet in this PR and have just some high level questions based on the comment of pre-read you added:
> 
> * if we are doing this ls -R or recursive readdir like resnet does initially from every thread before reading any data, are we going to trigger this pre-read? (that would be disastrous). i don't think the file is opened in this case, but not sure.

ls -R will not call open so no, there would be no difference to that code path.

> * in the case of a pre-read during the open, do we read 1 mib or up to the size of the file? is this querying the file size, or does fuse or dfuse already have queried the file size before issuing the open?

The file size is known during open and returned to the kernel.  This is then used for the read size.  The open performance should be unaffected (and there's no I/O changes during open).

> * is there a practical application this is aimed for?

This was tested with the Resnet training stage and helped, it should be useful for any dataset ingest which is dominated by read of smaller files.

**Review by mchaarawi — DISMISSED** [review]
looks good to me. im not clear though what happens in the case if the file was written to after the pre-read happens.  Is the pre-read buffer dropped?

**Comment by ashleypittman** [comment]
> looks good to me. im not clear though what happens in the case if the file was written to after the pre-read happens. Is the pre-read buffer dropped?

If the pre-read buffer is used then once it's exhausted (which might take several reads depending on application I/O sizes) then it will be released.  If there is other I/O, either writes or reads from elsewhere in the file then the buffer will not be used, but it won't be released until close.

File: src/client/dfuse/ops/read.c [hunk 1]
```diff
@@ -51,6 +51,55 @@ dfuse_cb_read_complete(struct dfuse_event *ev)
...
+		return true;
+	}
+
+	if (!oh->doh_linear_read || oh->doh_readahead->dra_ev == NULL) {
+		DFUSE_TRA_DEBUG(oh, "Readahead disabled");
+		return false;
+	}
+
+	if (oh->doh_linear_read_pos != position) {
+		DFUSE_TRA_DEBUG(oh, "disabling readahead");
```
Comment (daltonbohning) [inline src/client/dfuse/ops/read.c:70]: So "linear read" is synonymous with "readahead"?

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfuse/ops/read.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/read.c:70]: Apparently yes.  This PR started are read-ahead after the first read before I realised that the speculative read could happen on open.  This is debugging so I think it's Ok though.

Comment (daltonbohning) [reply src/client/dfuse/ops/read.c:70]: I was thinking from the perspective of the variable names, etc., but not asking you to change this

Comment (ashleypittman) [reply src/client/dfuse/ops/read.c:70]: https://github.com/daos-stack/daos/blob/master/src/client/dfuse/dfuse.h#L117-L126

`linear read` means the file was opened, read from start to end without seeking and closed.  We already track this to short-circuit the last zero-length read so we use that data to know if to use the pre-read buffer.

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**